scipy/stats/_hypotests.py

```diff
@@ -663,3 +664,329 @@ def somersd(x, y=None):
+    p-value of 6.4% :
+
+    >>> _, pvalue = stats.fisher_exact([[7, 12], [8, 3]], alternative="less")
+    >>> pvalue
+    0.0640...
+
+    With the same threshold probability of 5%, `fisher_exact` is accepting
+    :math:`H_0` while `barnard_exact` is rejecting it. As stated in [2]_,
+    Barnard's test is uniformly more powerful than Fisher's exact test
+    because Barnard's test do not condition on any margin. Fisher's test
+    should only be used when both sets of marginals are fixed.
+
+    """
+    if n_iter <= 0:
+        raise ValueError(
+            "Number of iterations `num_it` must be strictly positive, "
+            f"found {n_iter!r}"
+        )
+
+    table = np.asarray(table, dtype=np.int64)
+
+    if not table.shape == (2, 2):
+        raise ValueError("The input `table` must be of shape (2, 2).")
+
+    if np.any(table < 0):
+        raise ValueError("All values in `table` must be nonnegative.")
+
+    if 0 in table.sum(axis=0):
+        # If both values in column are zero, the p-value is 1 and
+        # the score's statistic is NaN.
+        return BarnardExactResult(np.nan, 1.0)
+
+    total_col_1, total_col_2 = table.sum(axis=0)
+
+    x1 = np.arange(total_col_1 + 1, dtype=np.int64).reshape(-1, 1)
+    x2 = np.arange(total_col_2 + 1, dtype=np.int64).reshape(1, -1)
+
+    # We need to calculate the wald statistics for each combination of x1 and
+    # x2.
+    p1, p2 = x1 / total_col_1, x2 / total_col_2
+
+    if pooled:
+        p = (x1 + x2) / (total_col_1 + total_col_2)
+        variances = p * (1 - p) * (1 / total_col_1 + 1 / total_col_2)
+    else:
+        variances = p1 * (1 - p1) / total_col_1 + p2 * (1 - p2) / total_col_2
+
+    # To avoid warning when dividing by 0
+    with np.errstate(divide="ignore", invalid="ignore"):
+        wald_statistic = np.divide((p1 - p2), np.sqrt(variances))
+
+    wald_statistic[p1 == p2] = 0  # Removing NaN values
+
+    wald_stat_obs = wald_statistic[table[0, 0], table[0, 1]]
+
+    if alternative == "two-sided":
+        index_arr = np.abs(wald_statistic) >= abs(wald_stat_obs)
+    elif alternative == "less":
+        index_arr = wald_statistic <= wald_stat_obs
+    elif alternative == "greater":
+        index_arr = wald_statistic >= wald_stat_obs
+    else:
+        msg = (
+            "`alternative` should be one of {'two-sided', 'less', 'greater'},"
+            f" found {alternative!r}"
+        )
+        raise ValueError(msg)
+
+    get_pvalue_from = _binomial_maximisation_of_p_value_with_nuisance_param(
+        total_col_1, total_col_2, index_arr
+    )
+
+    result = shgo(
+        get_pvalue_from,
+        bounds=((0, 1),),
+        n=32,  # Need to be a power of two since it is used by sobol
+        sampling_method="sobol",
+        iters=n_iter,
+    )
+
+    p_value = np.clip(-result.fun, a_min=0, a_max=1)
+    return BarnardExactResult(wald_stat_obs, p_value)
+
+
+def _binomial_maximisation_of_p_value_with_nuisance_param(
```

Comment:
Personally I preferred what you did before: having the optimization within this function and returning the p-value instead of returning a function. But if other reviewers are ok with this ok. If you keep it like this, you will need to update the docstring and possibly the name then.

Reply:
Sorry, it's my fault for not having been clear, you do not need a closure, you can use the `args=` keyword of sgo to pass fixed parameters to you p-value computing function.

I'll suggest a patch.

```
In [1]: from scipy.optimize import shgo

In [2]: def f(var, fixed1, fixed2):
   ...:     print(var, fixed1, fixed2)
   ...:     return var**2

In [3]: shgo(f, args=("Fixed:x1_sum_x2_log_comb", "Fixed:x1_sum_x2"), bounds=[(-1,1)])
[-1.] Fixed:x1_sum_x2_log_comb Fixed:x1_sum_x2
[1.] Fixed:x1_sum_x2_log_comb Fixed:x1_sum_x2
[0.] Fixed:x1_sum_x2_log_comb Fixed:x1_sum_x2
[0.] Fixed:x1_sum_x2_log_comb Fixed:x1_sum_x2
[1.49011612e-08] Fixed:x1_sum_x2_log_comb Fixed:x1_sum_x2
```

Reply:
I was more thinking along something like 633c3a8a34e4c64f450c4f63e59bf26fef3d7675. 
This way we have : no higher order functions, and no locally defined functions either.
@tupui what do you think ?

Reply:
Yes either fixing like this or with partial looks better 👍

Reply:
Did it in faa370d147d5364b45c1b58f69a0b53195b587d3. Thank you both